Rate Limit Docs + more flexible custom rules (#1080)
webrecorder/browsertrix-crawler: docs/docs/user-guide/rate-limits.md

# Rate Limit / CAPTCHA Page Detection

When crawling, it is often possible to encounter rate limits, CAPTCHAs, or auth pages that are shown instead of valid content that the crawler is intending to capture.
These 'auth-walled' pages are often implemented for good reason, and bypassing such limits is outside of the scope of the crawler.

However, the crawler can *detect* such pages and avoid archiving 'bad'/undesirable data.

The crawler includes several options to help with this. 

When pages are flagged as rate limited, the pages are still loaded in the browser, but the content is not archived.
Instead, the pages may be queued to be retried at a later time or skipped altogether.

## Flagging Rate Limited / Auth-walled pages

The simplest mechanism for detection is by checking the status code. By default, the crawler now skips all pages that have a 403, 429, or 503 status code. This list can be customized with the `--rateLimitStatusCodes` setting, which accepts a list of status codes to always skip and retry.

> The crawler also includes a custom option `--rateLimitOnMatch`, which can take one more values in the form of `<regex>` or `<regex>:<status>`, e.g. `--rateLimitOnMatch <regex-1> --rateLimitOnMatch <regex-2>:<status>`. This setting allows for custom detection of CAPTCHAs and rate-limited pages based on page content, and optionally status code, so even 200 pages can be flagged as rate limited.

Reviewer: ```suggestion
The crawler also includes a custom option `--rateLimitOnMatch`, which can take additional values in the form of `<regex>` or `<regex>:<status>`, e.g. `--rateLimitOnMatch <regex-1> --rateLimitOnMatch <regex-2>:<status>`. This setting allows for custom detection of CAPTCHAs and rate-limited pages based on page content, and optionally status code, so even 200 pages can be flagged as rate limited.
```

Reviewer: We should at least briefly mention *what* the regex matches.

If a site provides a `Retry-After` header, for example with a 429 response, the header value is used instead of the global default set in `--rateLimitTimeout`.

## Rate Limit Retry Count

By default, rate limited pages are retried indefinitely, allowing the crawl to possibly complete but at a much slower pace. If the pages are blocked due to auth requirements it may be possible to add a browser profile and restart the crawl, allowing for previously flagged pages to be captured successfully.

However, it may be desirable to set a total number of retries for pages flagged as rate limited, which can be done
by setting `--rateLimitMaxRetries` to a value >= 0, where 0 implies no retries at all.

## Interrupting Crawl on Rate Limit Threshold

By default, the crawler will continue, skipping rate limited pages and retrying them indefinitely, to match existing behavior.

If the `--rateLimitInterruptCount M` flag is set, the crawler will exit with a rate limit exit code (exit code 21) after M rate limited pages within the N seconds, configured via `--rateLimitTimeout`.

> This can allow another application (e.g. Kubernetes) to provide an exponential backoff system when the crawler is repeatedly exiting due to rate limited pages.

Reviewer: Maybe clarify that this allows an application/job runner which is operating the crawler to provide such a backoff.

Additionally, if direct fetch reaches this threshold, further direct fetches will also be skipped until the timeout expires.

> If a page loads successfully, the rate limit counter is cleared.

Reviewer: Does this mean any page?

Author: yes, updated to say 'any'

## Caveats and Multi-Domain Crawling

> Interrupting the crawl when the rate limit threshold is reached is disabled by default as it may not always be desirable when crawling many seeds across different domains, where one site may be rate limited while others are not. Currently, the rate limits are applied globally and not per-domain.

Reviewer: Might be helpful to flip the order of this a bit to clarify:

```suggestion
Because rate limits are applied globally and not per-domain, and it may not always be desirable to interrupt a crawl containing many seeds across different domains, the crawl will not be automatically interrupted when the rate limit threshold is reached.
```

Author: Well, there is no interrupt theshold by default for this reason, but it will be interrupt when it is reached, can reword.

# Rate Limit / CAPTCHA Page Detection

When crawling, it is often possible to encounter rate limits, CAPTCHAs, or auth pages that are shown instead of valid content that the crawler is intending to capture.
These 'auth-walled' pages are often implemented for good reason, and bypassing such limits is outside of the scope of the crawler.

However, the crawler can *detect* such pages and avoid archiving 'bad'/undesirable data.

The crawler includes several options to help with this. 

When pages are flagged as rate limited, the pages are still loaded in the browser, but the content is not archived.
Instead, the pages may be queued to be retried at a later time or skipped altogether.

## Flagging Rate Limited / Auth-walled pages

The simplest mechanism for detection is by checking the status code. By default, the crawler now skips all pages that have a 403, 429, or 503 status code. This list can be customized with the `--rateLimitStatusCodes` setting, which accepts a list of status codes to always skip and retry.

> The crawler also includes a custom option `--rateLimitOnMatch`, which can take additional values in the form of `<regex>` or `<regex>:<status>`, e.g. `--rateLimitOnMatch <regex-1> --rateLimitOnMatch <regex-2>:<status>`. This setting allows for custom detection of CAPTCHAs and rate-limited pages based on page content, and optionally status code, so even 200 pages can be flagged as rate limited.

Reviewer: Can we clarify what "page content" means in this context? I think it's helpful to be a bit more specific. I know what it means from reading the source, but the docs make it a little less clear. For example, we could specify the regular expression will be matched against the downloaded HTML/JavaScript.

Author: ```suggestion
The crawler also includes a custom option `--rateLimitOnMatch`, which can take additional values in the form of `<regex>` or `<regex>:<status>`, e.g. `--rateLimitOnMatch <regex-1> --rateLimitOnMatch <regex-2>:<status>`. This setting allows for custom detection of CAPTCHAs and rate-limited pages based on page text content (HTML, JS, etc...), and optionally status code, so even 200 pages can be flagged as rate limited.
```

How about this?

Reviewer: Looks good!

However, it may be desirable to set a total number of retries for pages flagged as rate limited, which can be done
by setting `--rateLimitMaxRetries` to a value >= 0, where 0 implies no retries at all.

## Interrupting Crawl on Rate Limit Threshold

By default, the crawler will continue, skipping rate limited pages and retrying them indefinitely, to match existing behavior.

If the `--rateLimitInterruptCount M` flag is set, the crawler will exit with a rate limit exit code (exit code 21) after M rate limited pages within the N seconds, configured via `--rateLimitTimeout`.

This can allow a job runner or other system that starts and monitors the crawler container to implement an exponential backoff system if the crawler repeatedly exits due to being rate limited.
The Browsertrix application provides this through built-in exponential backoff available in Kubernetes, where a crawler

> will restart only once every 5 minutes if continuously rate limited.

Reviewer 2: Not the biggest detail, but is the 5 minutes bit here accurate given that exponential backoff works by increasing the amount of time between restarts each time?

Author: Updated!
